File: keyboard/Sources/DocumentIdentityMonitor.swift
import Foundation

/// Detects rapid documentIdentifier churn (host input-session flapping, e.g.
/// WKWebView+React recreating input sessions). Pure logic: no UIKit, injectable
/// clock, fixed-capacity ring — observe() must not allocate (Jetsam hot path).
struct DocumentIdentityMonitor {
    /// One observed document identity at a moment in time.
    private struct Slot {
        var id: UUID?
        var ts: TimeInterval
    }

    /// Fixed ring capacity: at most 8 observations are considered per flap check.
    private let capacity = 8

    /// Flap = >=2 distinct non-nil ids OR >=2 nil<->non-nil transitions
    /// observed within flapWindow.
    let flapWindow: TimeInterval    // default 0.5 s
    let cooldown: TimeInterval      // defensive mode lasts this long after last flap; default 2.0 s
    let stability: TimeInterval     // early exit: same non-nil id held this long; default 0.5 s

    /// True while document identity churn is recent (host flapping).
    private(set) var isDefensive = false

    /// Ring of observations; oldest entry sits at nextIndex.
    private var slots: [Slot]
    /// Index of the next slot to overwrite (the oldest entry).
    private var nextIndex = 0
    /// Timestamp of the most recent flap (extended on re-flap).
    private var lastFlapAt: TimeInterval = 0
    /// When the current run of one identical non-nil id started.
    private var stableIDSince: TimeInterval?

    init(flapWindow: TimeInterval = 0.5, cooldown: TimeInterval = 2.0, stability: TimeInterval = 0.5) {
        self.flapWindow = flapWindow
        self.cooldown = cooldown
        self.stability = stability
        slots = [Slot](repeating: Slot(id: nil, ts: -.infinity), count: capacity)
    }

    /// Record one observation. Returns isDefensive after the update.
    mutating func observe(_ id: UUID?, at ts: TimeInterval) -> Bool {
        // The slot before the one being overwritten holds the previous observation.
        let previousID = slots[(nextIndex + capacity - 1) % capacity].id
        slots[nextIndex] = Slot(id: id, ts: ts)
        nextIndex = (nextIndex + 1) % capacity

        // Track the run of one identical non-nil id (breaks on nil or on any id
        // change) for the stability early exit.
        if id != nil {
            if stableIDSince == nil || previousID != id {
                stableIDSince = ts
            }
        } else {
            stableIDSince = nil
        }

        if detectsFlap(now: ts) {
            isDefensive = true
            lastFlapAt = ts
        }

        if isDefensive {
            if ts - lastFlapAt >= cooldown {
                isDefensive = false
            } else if id != nil, let since = stableIDSince, ts - since >= stability {
                isDefensive = false
            }
        }

        return isDefensive
    }

    /// True when the visible ring entries (those within flapWindow) show a flap:
    /// >=2 distinct non-nil ids or >=2 nil<->non-nil transitions.
    private func detectsFlap(now: TimeInterval) -> Bool {
        let windowStart = now - flapWindow

        // nil<->non-nil transitions between consecutive visible entries.
        var transitions = 0
        var previousWasNil: Bool?
        for i in 0..<capacity {
            let slot = slots[(nextIndex + i) % capacity]
            guard slot.ts >= windowStart else { continue }
            let isNil = (slot.id == nil)
            if let previous = previousWasNil, previous != isNil {
                transitions += 1
            }
            previousWasNil = isNil
        }
        if transitions >= 2 { return true }

        // Distinct non-nil ids among visible entries (O(n²) over <=8 entries,
        // UUID comparisons only — no allocation).
        var distinct = 0
        for i in 0..<capacity {
            let slotI = slots[(nextIndex + i) % capacity]
            guard let idI = slotI.id, slotI.ts >= windowStart else { continue }
            var seenBefore = false
            for j in 0..<i {
                let slotJ = slots[(nextIndex + j) % capacity]
                guard let idJ = slotJ.id, slotJ.ts >= windowStart else { continue }
                if idJ == idI {
                    seenBefore = true
                    break
                }
            }
            if !seenBefore {
                distinct += 1
            }
        }
        return distinct >= 2
    }
}

File: scripts/prediction-sim/lib/document-identity-monitor.mjs
// Pure-logic JS port of keyboard/Sources/DocumentIdentityMonitor.swift.
// Kept in sync per AGENTS.md -> Test policy.

/**
 * Detects rapid documentIdentifier churn (host input-session flapping, e.g.
 * WKWebView+React recreating input sessions). Pure logic, no UIKit; callers
 * supply their own clock via the `at` timestamp argument. Mirrors
 * DocumentIdentityMonitor in Swift: ids are strings (or null), timestamps are
 * caller-supplied numbers (fake clock).
 *
 * Flap = >=2 distinct non-nil ids OR >=2 nil<->non-nil transitions observed
 * within flapWindow. A flap sets isDefensive and records lastFlapAt; a new flap
 * while defensive extends the window. Defensive mode exits when cooldown has
 * elapsed since the last flap, or when the same non-nil id has been
 * continuously observed for >= stability (early exit).
 *
 * @param {number} [flapWindow=0.5] - Detection window in seconds.
 * @param {number} [cooldown=2.0] - Defensive mode lasts this long after the
 *   last flap.
 * @param {number} [stability=0.5] - Same non-nil id held this long exits early.
 */
export class DocumentIdentityMonitor {
  constructor(flapWindow = 0.5, cooldown = 2.0, stability = 0.5) {
    this.flapWindow = flapWindow;
    this.cooldown = cooldown;
    this.stability = stability;
    this.isDefensive = false;
    this.capacity = 8;
    // Fixed ring of slots; entries older than flapWindow are invisible.
    this.slots = [];
    for (let i = 0; i < this.capacity; i++) {
      this.slots.push({ id: null, ts: Number.NEGATIVE_INFINITY });
    }
    this.nextIndex = 0;
    this.lastFlapAt = 0;
    this.stableIDSince = null;
  }

  /**
   * Record one observation. Returns isDefensive after the update.
   * @param {string|null} id - Document identity, or null for none.
   * @param {number} at - Timestamp in seconds (caller-supplied fake clock).
   * @returns {boolean}
   */
  observe(id, at) {
    // The slot before the one being overwritten holds the previous observation.
    const previousID = this.slots[(this.nextIndex + this.capacity - 1) % this.capacity].id;
    this.slots[this.nextIndex] = { id, ts: at };
    this.nextIndex = (this.nextIndex + 1) % this.capacity;

    // Track the run of one identical non-nil id (breaks on null or on any id
    // change) for the stability early exit.
    if (id !== null) {
      if (this.stableIDSince === null || previousID !== id) {
        this.stableIDSince = at;
      }
    } else {
      this.stableIDSince = null;
    }

    if (this.detectsFlap(at)) {
      this.isDefensive = true;
      this.lastFlapAt = at;
    }

    if (this.isDefensive) {
      if (at - this.lastFlapAt >= this.cooldown) {
        this.isDefensive = false;
      } else if (id !== null && this.stableIDSince !== null && at - this.stableIDSince >= this.stability) {
        this.isDefensive = false;
      }
    }

    return this.isDefensive;
  }

  /**
   * True when the visible ring entries (those within flapWindow) show a flap:
   * >=2 distinct non-nil ids or >=2 nil<->non-nil transitions.
   * @param {number} now
   * @returns {boolean}
   */
  detectsFlap(now) {
    const windowStart = now - this.flapWindow;

    // nil<->non-nil transitions between consecutive visible entries.
    let transitions = 0;
    let previousWasNil = null;
    for (let i = 0; i < this.capacity; i++) {
      const slot = this.slots[(this.nextIndex + i) % this.capacity];
      if (slot.ts < windowStart) continue;
      const isNil = slot.id === null;
      if (previousWasNil !== null && previousWasNil !== isNil) transitions++;
      previousWasNil = isNil;
    }
    if (transitions >= 2) return true;

    // Distinct non-nil ids among visible entries (O(n²) over <=8 entries,
    // comparisons only).
    let distinct = 0;
    for (let i = 0; i < this.capacity; i++) {
      const slotI = this.slots[(this.nextIndex + i) % this.capacity];
      if (slotI.id === null || slotI.ts < windowStart) continue;
      let seenBefore = false;
      for (let j = 0; j < i; j++) {
        const slotJ = this.slots[(this.nextIndex + j) % this.capacity];
        if (slotJ.id === null || slotJ.ts < windowStart) continue;
        if (slotJ.id === slotI.id) {
          seenBefore = true;
          break;
        }
      }
      if (!seenBefore) distinct++;
    }
    return distinct >= 2;
  }
}
